### PR TITLE
feat: add redis adapter

### PR DESCRIPTION
**Pull request by kid-icarus**
Adding this for review. I've tested this locally and it seems to work just fine. Ultimately I'd like to add some integration tests here, as it has none, similar to the nodefs storage adapter.

For that to work, I was considering leveraging a [test container](https://testcontainers.com/guides/getting-started-with-testcontainers-for-nodejs/) to ensure that Redis is installed, but that would require an update to our test github action (to install docker).

I meant to open this PR earlier but thought I'd get a conversation started :)

### CODE REVIEW

**Comment by acurrieclark** [comment]
This looks good to me, but the only reason we know the existing adapters work is because we are using them in production!

As you mention, there aren't any specific tests for any of the Storage Adapters, and it would in indeed be good to have a standard set for all of them.

Without being in a position to test it myself, I am loathe to merge just yet. I imagine this will be something people are looking for though.

**Comment by pvh** [comment]
I think as the system matures we ought to start putting the storage & network adapters in their own repos (though then we lose automated test integration... hmm) but this is definitely something I've wanted to see for a while so thanks for sharing it!

File: packages/automerge-repo-storage-redis/LICENSE [hunk 1]
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2024-Present, Ryan Kois
```
Comment (bijela-gora) [inline packages/automerge-repo-storage-redis/LICENSE:1]: Looks like in this file you missed the first line. Like [here](https://github.com/facebook/react/blob/64f354cf27a17c4e5b0dfcd908e47940cf947771/LICENSE#L1C1-L1C12).

File: packages/automerge-repo-storage-redis/package.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  "type": "module",
+  "main": "dist/index.js",
+  "scripts": {
+    "build": "tsc",
+    "watch": "tsc -w"
+  },
+  "dependencies": {
+    "@automerge/automerge-repo": "workspace:*",
+    "debug": "^4.3.4",
+    "ioredis": "^5.3.2"
```
Comment (bijela-gora) [inline packages/automerge-repo-storage-redis/package.json:17]: It feels like the `ioredis` must be peer dependency. What do you think?

File: packages/automerge-repo-storage-redis/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+export class RedisStorageAdapter implements StorageAdapterInterface {
+  private redis: Redis
+
+  /**
+   * Create a new {@link RedisStorageAdapter}.
+   * @param opts - Options to pass to the Redis client.
+   */
+  constructor(
+    opts?: RedisOptions
```
Comment (bijela-gora) [inline packages/automerge-repo-storage-redis/src/index.ts:26]: Maybe Redis instance must be passed as a constructor parameter?

File: packages/automerge-repo-storage-redis/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+  async remove(keyArray: string[]): Promise<void> {
+    await this.redis.del(keyArray.join(":"))
+  }
+
+  async loadRange(keyPrefix: string[]): Promise<Chunk[]> {
+    const lowerBound = [...keyPrefix, '*'].join(":")
+    const result: Chunk[] = []
+    const stream = this.redis.scanStream({
```
Comment (bijela-gora) [inline packages/automerge-repo-storage-redis/src/index.ts:52]: Nice!

**Review by bijela-gora — COMMENTED** [review]
Hi,

I decided to check the list of PRs before trying to code a Redis-based implementation of `StorageAdapterInterface`. And here it is. Thank you for your work! Don't you mind I will review it?

**Comment by bijela-gora** [comment]
@pvh hello,

Your phrase: "though then we lose automated test integration... hmm" reminded me the [Promises/A+ Compliance Test Suite](https://github.com/promises-aplus/promises-tests). What do you think of the idea of developing a set of tests that could be used to test any implementation of the `StorageAdapterInterface`?

**Comment by pvh** [comment]
Sounds great if it can be kept simple. I think the tricky part here is operational complexity -- for things like nodefs & indexeddb there are easily available implementations but for Redis, Postgres, or other backends, we'll need to have the ability to spin up the service at runtime. For SaaS-only backends (say, S3) there's yet more complexity.

That said, I'm quite interested in a solution. I think @HerbCaudill was the one who designed the system we use to test the network adapters and as we grow beyond a monorepo, I think something along the lines you propose could be valuable.

File: packages/automerge-repo-storage-redis/package.json (hunk 2 above)
Comment (kid-icarus) [reply packages/automerge-repo-storage-redis/package.json:17]: Yeah that makes sense, I agree

**Comment by bijela-gora** [comment]
@kid-i

**Comment by pvh** [comment]
@kid-icarus How do you feel about hosting this as a standalone package at /automerge/automerge-repo-storage-redis/?

I think my intuition is that adapters which rely on things not found in most browsers / OS should have their own repo and then whatever CI we set up can be independent.

**Comment by bijela-gora** [comment]
@pvh

I'd like to share that I created an SQLite storage adapter for JS/TS projects: https://github.com/bijela-gora/automerge-repo-storage-better-sqlite3

I created it because I'm going to use it in my pet project.

Thank you for the awesome Automerge project!

**Comment by kid-icarus** [comment]
> @kid-icarus How do you feel about hosting this as a standalone package at /automerge/automerge-repo-storage-redis/?
> 
> I think my intuition is that adapters which rely on things not found in most browsers / OS should have their own repo and then whatever CI we set up can be independent.

That makes sense from a principled perspective 👍🏻

One pro of a separate repo is more control over CI, being able to throw whatever external dependencies into a Dockerfile, etc. Another pro is that if there are breaking changes in automerge(-repo), namely the types in this specific case, the burden isn't placed on the core team to update the adapter in the monorepo.

One con would be extra care to specify automerge and automerge-repo as a combination of dev and peer dependencies, so that adapter consumers aren't using duplicate, potentially incompatible versions of them. This also means that as newer versions of automerge(-repo) are released, we'll need to bump those dev dependencies. Dependabot can help with that, but it's just the general con of something living in a monorepo vs an external repo.

**Comment by akellbl4** [comment]
Not sure how far `ioredis` from `node-redis`, but here is [a part of the description](https://github.com/redis/ioredis#readme) from `ioredis` package.

> For new projects, node-redis is the recommended client library. [node-redis](https://github.com/redis/node-redis) is the open-source (MIT license) Redis JavaScript client library redesigned from the ground up and actively maintained. [node-redis](https://github.com/redis/node-redis) supports new (hash-field expiration) and future commands and the capabilities available in Redis Stack and Redis 8 (search, JSON, time-series, probabilistic data structures).

I guess if API lines up well enough storage can allow passing an instance of both `ioredis` and `node-redis`
